apache/incubator-yunikorn-k8shim
# [YUNIKORN-686] [Regression] Placeholders for completed applications are recreated during recovery

### What is this PR for?
After https://issues.apache.org/jira/browse/YUNIKORN-677, @yuchaoran2011 found a regression. See more info in https://issues.apache.org/jira/browse/YUNIKORN-686. The fix in #265 introduces an issue that while we do recovery, we added the terminated pod into the app with New state, this will cause the scheduler to try to schedule the pod again. When GS is enabled, the placeholders are created again.

While fixing this issue, I decided to do some code refactoring to simplify the logic. That's why it seems I have changed more lines of code than expected. The refactoring removes the "Recovery" flag in `TaskMetadata`, and moves the logic to `task. initialization()`, this ensures when we create the task in the cache, we always ensure the task state is initialized correctly. To be more specific:

1.  if the task pod was already allocated and running, the task state post-recovery will be `Allocated`
2. if the task pod was already allocated and terminated (failed or succeed), the task state post-recovery will be "Completed"
3. if the task pod was in Pending state, the task state post-recovery will be `New`

Also, this PR includes some new UT to cover these scenarios.

### What type of PR is it?

### Todos

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-686

### How should this be tested?
Tested recovery locally:
1.  submit a job and let the job finish, restart the scheduler, make sure the app was recovered with 0 allocation
2. submit a job and keep the job running, restart the scheduler, make sure the app was recovered with all existing allocations recovered
3. submit a GS job, wait for the job to complete. restart the scheduler, make sure no placeholders are created

### Screenshots (if appropriate)

### Questions:
